fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/ResourceFactoryImpl.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.impl.models;
+
+import org.fcrepo.kernel.api.exception.InteractionModelViolationException;
+import org.fcrepo.kernel.api.models.Container;
+import org.fcrepo.kernel.api.models.FedoraBinary;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.FedoraTimeMap;
+import org.fcrepo.kernel.api.models.FedoraWebacAcl;
+import org.fcrepo.kernel.api.models.NonRdfSourceDescription;
+import org.fcrepo.kernel.api.models.ResourceFactory;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionFactory;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+
+
+/**
+ * Implementation of ResourceFactory interface.
+ *
+ * @author whikloj
+ * @since 2019-09-23
+ */
+public class ResourceFactoryImpl implements ResourceFactory {
+
+    private final PersistentStorageSession psSession;
+
+    /**
+     * Special constructor for testing.
+     *
+     * @param psFactory Persistent storage session factory.
+     */
+    protected ResourceFactoryImpl(final PersistentStorageSessionFactory psFactory) {
+        this.psSession = psFactory.getReadOnlySession();
+    }
+
+    @Override
+    public Container findOrInitContainer(final String identifier, final String interactionModel) {
+        final FedoraResource container = findInPersistence(identifier);
+        if (!container.getTypes().stream().anyMatch(t -> interactionModel.equalsIgnoreCase(t.toString()))) {
```

Comment:
getTypes would return all of the types assigned to a resource, wouldn't it? So there would be user assigned types. Maybe we need a `getInteractionModel(s)` method?
In terms of interaction models, are there any cases where two interaction models can be on the same resource?

Reply:
I may be unclear on what the definition of an interaction model is. The LDP spec makes limited use of it as well.

We can assign both DirectContainer and RDFSource "types" to an object, and seemingly these are both interaction models but really I think the DirectContainer is the only interaction model.